### ISSUE TITLE
Runtime config metrics sinks configuration seem broken

### ISSUE THREAD

**Issue by ChristianEder** [tomkerkhove/promitor#1052]
My goal is to configure promitor to only report metrics via its own /metrics scraping endpoint for prometheus. I can see in the [docs](https://github.com/tomkerkhove/promitor/blob/master/docs/configuration/v1.x/runtime.md#prometheus-scraping-endpoint---legacy-configuration) that this configuratio has been changed in version 1.6 (which I am using). However, I cannot get this to work. 

## Expected Behavior
I can configure Promitor to only use the built-in prometheus scraping endpoint metric sink.

## Actual Behavior
I tried 3 different configurations:

1) Use only the new "metrics sinks" configuration:

```yaml
server:
  httpPort: 80
metricSinks:
  prometheusScrapingEndpoint:
    metricUnavailableValue: NaN
    enableMetricTimestamps: true
    baseUriPath: /metrics
metricsConfiguration:
  absolutePath: /config/metrics-declaration.yaml
telemetry:
  applicationInsights:
    isEnabled: false
    verbosity: trace
  containerLogs:
    isEnabled: true
  defaultVerbosity: "Error"
```

with this configuration, I get an exception on startup:

```
██████╗ ██████╗  ██████╗ ███╗   ███╗██╗████████╗ ██████╗ ██████╗
██╔══██╗██╔══██╗██╔═══██╗████╗ ████║██║╚══██╔══╝██╔═══██╗██╔══██╗
██████╔╝██████╔╝██║   ██║██╔████╔██║██║   ██║   ██║   ██║██████╔╝
██╔═══╝ ██╔══██╗██║   ██║██║╚██╔╝██║██║   ██║   ██║   ██║██╔══██╗
██║     ██║  ██║╚██████╔╝██║ ╚═╝ ██║██║   ██║   ╚██████╔╝██║  ██║
╚═╝     ╚═╝  ╚═╝ ╚═════╝ ╚═╝     ╚═╝╚═╝   ╚═╝    ╚═════╝ ╚═╝  ╚═╝
[05:48:50 FTL] Host terminated unexpectedly
System.NullReferenceException: Object reference not set to an instance of an object.
   at Promitor.Agents.Scraper.Startup..ctor(IConfiguration configuration) in /src/Promitor.Agents.Scraper/Startup.cs:line 33
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.UseStartup(Type startupType, HostBuilderContext context, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass12_0.<UseStartup>b__0(HostBuilderContext context, IServiceCollection services)
   at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
   at Promitor.Agents.Scraper.Program.Main(String[] args) in /src/Promitor.Agents.Scraper/Program.cs:line 23
```

2) Use only the deprecated configuration

```yaml
server:
  httpPort: 80
prometheus:
  metricUnavailableValue: NaN # Optional. Default: NaN
  enableMetricTimestamps: false # Optional. Default: true
  scrapeEndpoint:
    baseUriPath: /metrics # Optional. Default: /metrics
metricsConfiguration:
  absolutePath: /config/metrics-declaration.yaml
telemetry:
  applicationInsights:
    isEnabled: false
    verbosity: trace
  containerLogs:
    isEnabled: true
  defaultVerbosity: "Error"
```
With this, I get a NullReferenceException on startup which seems to indicate (when looking at the source code) that the metrics sinks configuration secgtion is missing (which it is indeed):

```
██████╗ ██████╗  ██████╗ ███╗   ███╗██╗████████╗ ██████╗ ██████╗ 
██╔══██╗██╔══██╗██╔═══██╗████╗ ████║██║╚══██╔══╝██╔═══██╗██╔══██╗
██████╔╝██████╔╝██║   ██║██╔████╔██║██║   ██║   ██║   ██║██████╔╝
██║     ██║  ██║╚██████╔╝██║ ╚═╝ ██║██║   ██║   ╚██████╔╝██║  ██║
╚═╝     ╚═╝  ╚═╝ ╚═════╝ ╚═╝     ╚═╝╚═╝   ╚═╝    ╚═════╝ ╚═╝  ╚═╝
[05:51:33 FTL] Host terminated unexpectedly
System.NullReferenceException: Object reference not set to an instance of an object.
   at Promitor.Agents.Scraper.Startup.BuildOpenApiDescription(IConfiguration configuration) in /src/Promitor.Agents.Scraper/Startup.cs:line 111
   at Promitor.Agents.Scraper.Startup.ConfigureServices(IServiceCollection services) in /src/Promitor.Agents.Scraper/Startup.cs:line 39
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.InvokeCore(Object instance, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass9_0.<Invoke>g__Startup|0(IServiceCollection serviceCollection)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.Invoke(Object instance, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass8_0.<Build>b__0(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.UseStartup(Type startupType, HostBuilderContext context, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass12_0.<UseStartup>b__0(HostBuilderContext context, IServiceCollection services)
   at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
   at Promitor.Agents.Scraper.Program.Main(String[] args) in /src/Promitor.Agents.Scraper/Program.cs:line 23
```

3) Use both the deprecated and the new configuration

```yaml
server:
  httpPort: 80
metricSinks:
  prometheusScrapingEndpoint:
    metricUnavailableValue: NaN
    enableMetricTimestamps: true
    baseUriPath: /metrics
prometheus:
  metricUnavailableValue: NaN # Optional. Default: NaN
  enableMetricTimestamps: false # Optional. Default: true
  scrapeEndpoint:
    baseUriPath: /metrics # Optional. Default: /metrics
metricsConfiguration:
  absolutePath: /config/metrics-declaration.yaml
telemetry:
  applicationInsights:
    isEnabled: false
    verbosity: trace
  containerLogs:
    isEnabled: true
  defaultVerbosity: "Error"
```

With this, I get a validation error indicating that I specified duplicate metric sinks, which I did indeed:

```
[05:53:55 INF] Validation step 3/5 succeeded
[05:53:55 INF] Start Validation step 4/5: StatsD Metric Sink
[05:53:55 INF] Validation step 4/5 succeeded
[05:53:55 INF] Start Validation step 5/5: Prometheus Scraping Endpoint Metric Sink
[05:53:55 WRN] Validation step 5/5 failed. Error(s): Duplicate Prometheus scraping endpoint was configured in the metric sink & legacy configuration
[05:53:55 FTL] Promitor is not configured correctly. Please fix validation issues and re-run.
[05:53:55 FTL] Host terminated unexpectedly
Promitor.Agents.Scraper.Validation.Exceptions.ValidationFailedException: Validation Failed. Errors:- Prometheus Scraping Endpoint Metric Sink: Duplicate Prometheus scraping endpoint was configured in the metric sink & legacy configuration

   at Promitor.Agents.Scraper.Validation.RuntimeValidator.ProcessValidationResults(List`1 validationResults) in /src/Promitor.Agents.Scraper/RuntimeValidator.cs:line 60   at Promitor.Agents.Scraper.Validation.RuntimeValidator.Run() in /src/Promitor.Agents.Scraper/RuntimeValidator.cs:line 47
   at Promitor.Agents.Scraper.Startup.ValidateRuntimeConfiguration(IServiceCollection services) in /src/Promitor.Agents.Scraper/Startup.cs:line 86
   at Promitor.Agents.Scraper.Startup.ConfigureServices(IServiceCollection services) in /src/Promitor.Agents.Scraper/Startup.cs:line 51
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.InvokeCore(Object instance, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass9_0.<Invoke>g__Startup|0(IServiceCollection serviceCollection)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.Invoke(Object instance, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass8_0.<Build>b__0(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.UseStartup(Type startupType, HostBuilderContext context, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass12_0.<UseStartup>b__0(HostBuilderContext context, IServiceCollection services)
   at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
   at Promitor.Agents.Scraper.Program.Main(String[] args) in /src/Promitor.Agents.Scraper/Program.cs:line 23
```

The only way I got this working at all was to specify the prometheus scraping endpoint using the legacy configuration, and specify ANOTHER sink (statsd) using the new metricsSinks configuration.

## Steps to Reproduce the Problem

See actual behaviour.

## Specifications

  - Version: 1.6.0
  - Platform: Windows 10 Host, using Linux Containers

**Comment by tomkerkhove** [comment]
Hm odd, option one and two worked for me. Will have a look and sorry for the inconvenience..

**Comment by tomkerkhove** [comment]
I did some changes this week, would you mind trying the latest tag on our CI image on Docker Hub?

**Comment by ChristianEder** [comment]
Could you share a specific tag I should try it with?

**Comment by ChristianEder** [comment]
Or did you literally mean the "latest" tag?

**Comment by ChristianEder** [comment]
so this one? https://hub.docker.com/layers/tomkerkhove/promitor-agent-scraper/latest/images/sha256-788f1817bed19e708538afc3151c57f32d7d81c10f470965fb1c385570607ca8?context=explore

**Comment by ChristianEder** [comment]
because with that version, I get the error. My exact setup:

metrics-config.yaml:
```yaml
version: v1
azureMetadata:
  tenantId: mytenant
  subscriptionId: mysubscription
  resourceGroupName: myresourcegroup
metricDefaults:
  aggregation:
    interval: 00:05:00
  scraping:
    schedule: "0 */2 * ? * *"
  labels:
    metrics_owner_component: "some label"
    metrics_owner_servicename: "another label"
metrics:
- azureMetricConfiguration:
    aggregation:
      type: Maximum
    metricName: dailyMessageQuotaUsed
  description: The total number of messages used on the current day
  name: azure_iothub_daily_message_quota
  resourceType: Generic
  resources:
  - resourceUri: Microsoft.Devices/IotHubs/somehub
```

runtime-config.yaml:
```yaml
server:
  httpPort: 80
metricSinks:
  prometheusScrapingEndpoint:
    metricUnavailableValue: NaN
    enableMetricTimestamps: true
    baseUriPath: /metrics
metricsConfiguration:
  absolutePath: /config/metrics-declaration.yaml
telemetry:
  applicationInsights:
    isEnabled: false
    verbosity: trace
  containerLogs:
    isEnabled: true
  defaultVerbosity: "Error"
```

Startup script:
```
docker run -d -p 8999:80 --name promitor-agent-scraper --env PROMITOR_AUTH_APPID='myappid' --env PROMITOR_AUTH_APPKEY='myappkey' --volume C:/path/to/metrics-config.yaml:/config/metrics-declaration.yaml --volume C:/path/to/runtime-config.yaml:/config/runtime.yaml tomkerkhove/promitor-agent-scraper:latest
```

and I verrifed I got the right version as latest:
```
docker inspect --format='{{index .RepoDigests 0}}' tomkerkhove/promitor-agent-scraper:latest

tomkerkhove/promitor-agent-scraper@sha256:788f1817bed19e708538afc3151c57f32d7d81c10f470965fb1c385570607ca8
```

**Comment by tomkerkhove** [comment]
I was more referring to this one:
`> docker pull tomkerkhove/promitor-agent-scraper-ci:pr1051`

But I think there is another bug that I'm fixing now :(

**Comment by ChristianEder** [comment]
Hmm with this image I get the "Duplicate Prometheus scraping endpoint was configured in the metric sink & legacy configuration" error even though I only have configured the new metric sink version.

**Comment by tomkerkhove** [comment]
Yes, I'm fixing that and shipping 1.6.1 since there is a default value messing things up which I didn't use in testing - Sorry!

**Comment by tomkerkhove** [comment]
Fixed and will be shipped in v1.6.1 but can use this image for now:
> docker pull tomkerkhove/promitor-agent-scraper-ci:pr1054-linux

**Comment by ChristianEder** [comment]
that one fixes it, yes
